drivers/overlay/ostree.go

```diff
@@ -0,0 +1,122 @@
+// +build ostree
+
+package overlay
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"github.com/containers/storage/pkg/idtools"
+	"github.com/containers/storage/pkg/system"
+	"github.com/ostreedev/ostree-go/pkg/otbuiltin"
+	"github.com/pkg/errors"
+)
+
+func ostreeSupport() bool {
+	return true
+}
+
+func fixFiles(dir string, usermode bool) error {
+	entries, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return err
+	}
+
+	for _, info := range entries {
+		fullpath := filepath.Join(dir, info.Name())
+		if info.Mode()&(os.ModeNamedPipe|os.ModeSocket|os.ModeDevice) != 0 {
+			if err := os.Remove(fullpath); err != nil {
+				return err
+			}
+			continue
+		}
+
+		if info.IsDir() {
+			if usermode {
+				if err := os.Chmod(fullpath, info.Mode()|0700); err != nil {
+					return err
+				}
+			}
+			err = fixFiles(fullpath, usermode)
+			if err != nil {
+				return err
+			}
+		} else if usermode && (info.Mode().IsRegular()) {
+			if err := os.Chmod(fullpath, info.Mode()|0600); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// Create prepares the filesystem for the OSTREE driver and copies the directory for the given id under the parent.
+func convertToOSTree(repoLocation, root, id string) error {
+	repo, err := otbuiltin.OpenRepo(repoLocation)
+	if err != nil {
+		return errors.Wrap(err, "could not open the OSTree repository")
+	}
+
+	if _, err := repo.PrepareTransaction(); err != nil {
+		return errors.Wrap(err, "could not prepare the OSTree transaction")
+	}
+
+	if err = fixFiles(root, os.Getuid() != 0); err != nil {
+		return errors.Wrap(err, "could not prepare the OSTree directory")
+	}
+
+	commitOpts := otbuiltin.NewCommitOptions()
+	commitOpts.Timestamp = time.Now()
+	commitOpts.LinkCheckoutSpeedup = true
+	commitOpts.Parent = "0000000000000000000000000000000000000000000000000000000000000000"
+	branch := fmt.Sprintf("containers-storage/%s", id)
+
+	if _, err := repo.Commit(root, branch, commitOpts); err != nil {
+		return errors.Wrap(err, "FAIL could not commit the layer")
+	}
+
+	if _, err := repo.CommitTransaction(); err != nil {
+		return errors.Wrap(err, "could not complete the OSTree transaction")
+	}
+
+	if err := system.EnsureRemoveAll(root); err != nil {
+		return err
+	}
+
+	checkoutOpts := otbuiltin.NewCheckoutOptions()
+	checkoutOpts.RequireHardlinks = true
+	checkoutOpts.Whiteouts = true
```

Comment:
Whiteout information is stored differently for overlay than it is by aufs, and the format is converted when creating and extracting archives using pkg/archives.  What does this flag cause ostree to do?

Reply:
it tells ostree to handle files with a prefix ".wh." (i.e. delete the existing file/directory) in the checkout directory.

Reply:
That's how aufs stores whiteouts (and I'm not sure that what you're describing handles opaque directories), but in overlay whiteouts are character device nodes with major/minor of 0/0.  (Under [whiteouts and opaque directories](https://www.kernel.org/doc/Documentation/filesystems/overlayfs.txt) in the overlayfs docs.)

Reply:
@cgwalters Does ostree handle opaque directory whiteouts?  I took a quick look at `ostree-repo-checkout.c` and I couldn't find logic specific to handling entries named `.wh..wh..opq`.

Reply:
@cgwalters It appears not, then.

Reply:
I've opened a PR to support this: https://github.com/ostreedev/ostree/pull/1486

Reply:
Thanks for that.
If I'm reading it right, `fixFiles` removes file whiteouts from diff directories, and that's a problem.

Reply:
ah thanks to have found this issue.  I thought there were no whiteouts files, I've changed the code so that there is no special handling of these files.